<a href="https://colab.research.google.com/github/sumanthgm/Data-Science/blob/main/python/s5Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="faculty.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Faculty Notebook <br> (Session 5) </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

## Table of Content

1. **[Pandas](#pandas)**

2. **[Data Structures](#structures)**
    
3. **[Pandas Series](#series)**
    - 3.1 - [Creating a Series](#creatingS)
    - 3.2 - [Manipulating Series](#manipulatingS)

4. **[Pandas DataFrames](#dataframes)**
    - 4.1 - [Creating DataFrames](#creatingDF)
    - 4.2 - [Manipulating DataFrames](#manipulatingDF)

5. **[Reading Data from Different Sources](#reading_data)**


<a id="pandas"> </a>
## 1. Pandas

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> Pandas contain data structures and data manipulation tools designed for data cleaning and analysis.
<br><br>
                        While pandas adopt many coding idioms from NumPy, the biggest difference is that pandas is designed for working with tabular or heterogeneous data. NumPy, by contrast, is best suited for working with homogeneous numerical array data.<br><br>
                         The name Pandas is derived from the term “panel data”, an econometrics term for multidimensional structured data sets.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

**How to install pandas?**<br>
1. You can use-<br>
`!pip install pandas`<br>
2. You can import it as-<br>
import pandas as pd

To import the pandas library the following convention is used

In [3]:
import pandas as pd

So from now on we will use `pd.` instead of pandas.

<a id="structures"> </a>
## 2. Data Structures

Pandas has two data structures as follows:<br>
1. A Series is 1-dimensional labeled array that can hold data of any type (integer, string, boolean, float, python objects, and so on). It’s axis labels are collectively called an index.<br>
2. A DataFrame is 2-dimensional labeled data structure with columns. It supports multiple datatypes.
  

<a id="series"> </a>
## 3. Pandas Series

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> Pandas Series is a one-dimensional labeled array capable of holding any data type. However, a series is a sequence of homogeneous data types, similar to an array, list, or column in a table.<br><br>
                        It will assign a labeled index to each item in the Series. By default, each item will receive an index label from 0 to N, where N is the length of the Series minus one.
<br><br>
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

<a id="creatingS"> </a>
### 3.1 Creating a Series

**1. To create a numeric series**

In [4]:
# create a numeric series
numbers = range(1,100,5)
pd.Series(numbers)

0      1
1      6
2     11
3     16
4     21
5     26
6     31
7     36
8     41
9     46
10    51
11    56
12    61
13    66
14    71
15    76
16    81
17    86
18    91
19    96
dtype: int64

The output also gives the data type of the series as `int64`

And note that by default, each item will receive an index label from 0 to N, where N is the length of the Series minus one.

<b>*In python, the row names are called 'index'*</b>

**2. To create an object series**

In [5]:
# create a object series
string = "Hi" , "How" ,"are", "you", "?"
pd.Series(string)

0     Hi
1    How
2    are
3    you
4      ?
dtype: object

The output gives the data type of the series as `object`

**3. To create a series by giving both numeric and string values**

In [6]:
# create a Series with an arbitrary list
s = pd.Series([345, 'London', 34.5, -34.45, 'Happy Birthday'])
s

0               345
1            London
2              34.5
3            -34.45
4    Happy Birthday
dtype: object

Here the numeric values are treated as object.

**4. To set index values for a series**

In [7]:
marks = [60, 89, 74, 86]

subject = ["Maths", "Science", "English" , "Social Science"]

pd.Series(marks, index = subject)


Maths             60
Science           89
English           74
Social Science    86
dtype: int64

The index is added using the argument `index=`. The data type of the series continues to be numeric.

**5. To create a series from a dictionary**

In [8]:
data = {'Maths': 60, 'Science': 89, 'English': 76, 'Social Science': 86}

pd.Series(data)

Maths             60
Science           89
English           76
Social Science    86
dtype: int64

On passing a dict, the index in the resulting Series will have the dict’s keys in sorted order.

**6. To create a series using library `numpy`**

In [9]:
import numpy as np
s = np.linspace(0,10, 5)
s

array([ 0. ,  2.5,  5. ,  7.5, 10. ])

We see that we obtain an array. We can now convert it to a Series.

In [10]:
pd.Series(s)

0     0.0
1     2.5
2     5.0
3     7.5
4    10.0
dtype: float64

**7. A series with missing values**

If we pass a key that is not defined then its value will be `NAN`.

In [11]:
subjects = ["Maths", "Science", "Art and Craft" , "Social Science"]

marks_series = pd.Series(data, index = subjects)

print(marks_series)

Maths             60.0
Science           89.0
Art and Craft      NaN
Social Science    86.0
dtype: float64


<a id="manipulatingS"> </a>
### 3.2 Manipulating Series

**1. To check for null values using `.isnull`**

In [12]:
marks_series.isnull()

Maths             False
Science           False
Art and Craft      True
Social Science    False
dtype: bool

`False` indicates that the value is not null.

**2. To check for null values using `.notnull`**

In [13]:
marks_series.notnull()

Maths              True
Science            True
Art and Craft     False
Social Science     True
dtype: bool

` True` indicates that the value is not null.

**3. To know the subjects in which marks score is more than 75**

In [14]:
marks_series[marks_series > 75]

Science           89.0
Social Science    86.0
dtype: float64

**4. To assign 68 marks to 'Art and Craft'**

In [15]:
marks_series["Art and Craft"] = 68

In [17]:
marks_series

Maths             60.0
Science           89.0
Art and Craft     68.0
Social Science    86.0
dtype: float64

**5. To check whether Maths marks are 73**

In [18]:
marks_series.Maths == 73

False

In [19]:
# or you may use

marks_series["Maths"] == 73

False

**6. To create a series by generating numpy random numbers**

In [20]:
import numpy as np
num = pd.Series(np.random.random(15)*10).round()

**7. To find the square of the numbers series**

In [21]:
square = pd.Series(num*num)
square.index = [num]
square

1.0     1.0
9.0    81.0
6.0    36.0
7.0    49.0
9.0    81.0
2.0     4.0
9.0    81.0
5.0    25.0
6.0    36.0
3.0     9.0
3.0     9.0
2.0     4.0
6.0    36.0
8.0    64.0
7.0    49.0
dtype: float64

**8. To assign index name and object name**

In [22]:
square.name = 'Square'
square.index.name = 'Number'
square

1.0     1.0
9.0    81.0
6.0    36.0
7.0    49.0
9.0    81.0
2.0     4.0
9.0    81.0
5.0    25.0
6.0    36.0
3.0     9.0
3.0     9.0
2.0     4.0
6.0    36.0
8.0    64.0
7.0    49.0
Name: Square, dtype: float64

**Note:** A Series’s index can be altered in-place by assignment

From the output, it is not clear that the index column is labeled, to check whether it is labeled let us print it

In [23]:
square.index

MultiIndex([(1.0,),
            (9.0,),
            (6.0,),
            (7.0,),
            (9.0,),
            (2.0,),
            (9.0,),
            (5.0,),
            (6.0,),
            (3.0,),
            (3.0,),
            (2.0,),
            (6.0,),
            (8.0,),
            (7.0,)],
           name='Number')

From `name='Number'`, it is seen that the column is labeled.

**9. Add a number 5 to every element of the series**

In [24]:
square + 5

1.0     6.0
9.0    86.0
6.0    41.0
7.0    54.0
9.0    86.0
2.0     9.0
9.0    86.0
5.0    30.0
6.0    41.0
3.0    14.0
3.0    14.0
2.0     9.0
6.0    41.0
8.0    69.0
7.0    54.0
Name: Square, dtype: float64

The number 5 has been added to all the variables.

**10. To extract a value specifying the index**

In [25]:
square[6]

6.0    36.0
6.0    36.0
6.0    36.0
Name: Square, dtype: float64

All values with index as '6' are obtained

**11. To extract a range of values specifying the location**

In [26]:
square[6:14]

9.0    81.0
5.0    25.0
6.0    36.0
3.0     9.0
3.0     9.0
2.0     4.0
6.0    36.0
8.0    64.0
Name: Square, dtype: float64

**12. Usage of `.iloc`**

We use `.iloc` to get the values of the specified index of numbers

In [27]:
square.iloc[3]

49.0

In [28]:
square.iloc[4:9]

9.0    81.0
2.0     4.0
9.0    81.0
5.0    25.0
6.0    36.0
Name: Square, dtype: float64

**13. Sorting a numeric series**

In [29]:
# create a pandas series
values = pd.Series([23, 45, np.nan, 41, 23, 34, 55, np.nan, 34, 20])

values

0    23.0
1    45.0
2     NaN
3    41.0
4    23.0
5    34.0
6    55.0
7     NaN
8    34.0
9    20.0
dtype: float64

In [30]:
# ascending order
values.sort_values(ascending = True)

9    20.0
0    23.0
4    23.0
5    34.0
8    34.0
3    41.0
1    45.0
6    55.0
2     NaN
7     NaN
dtype: float64

In [31]:
# descending order
values.sort_values(ascending = False)

6    55.0
1    45.0
3    41.0
5    34.0
8    34.0
0    23.0
4    23.0
9    20.0
2     NaN
7     NaN
dtype: float64

**14. Sorting a categorical series**

In [32]:
# create a pandas series
string_values = pd.Series(["a", "j", "d", "f", "t", "a"])

string_values

0    a
1    j
2    d
3    f
4    t
5    a
dtype: object

In [33]:
# ascending order
string_values.sort_values(ascending = True)

0    a
5    a
2    d
3    f
1    j
4    t
dtype: object

In [34]:
# descending order
string_values.sort_values(ascending = False)

4    t
1    j
3    f
2    d
0    a
5    a
dtype: object

**15. Sorting based on index**

In [35]:
# recall the square
square

1.0     1.0
9.0    81.0
6.0    36.0
7.0    49.0
9.0    81.0
2.0     4.0
9.0    81.0
5.0    25.0
6.0    36.0
3.0     9.0
3.0     9.0
2.0     4.0
6.0    36.0
8.0    64.0
7.0    49.0
Name: Square, dtype: float64

In [36]:
# sort in ascending order based on index
square.sort_index(ascending = True)

1.0     1.0
2.0     4.0
2.0     4.0
3.0     9.0
3.0     9.0
5.0    25.0
6.0    36.0
6.0    36.0
6.0    36.0
7.0    49.0
7.0    49.0
8.0    64.0
9.0    81.0
9.0    81.0
9.0    81.0
Name: Square, dtype: float64

In [37]:
# sort in descending order based on index
square.sort_index(ascending = False)

9.0    81.0
9.0    81.0
9.0    81.0
8.0    64.0
7.0    49.0
7.0    49.0
6.0    36.0
6.0    36.0
6.0    36.0
5.0    25.0
3.0     9.0
3.0     9.0
2.0     4.0
2.0     4.0
1.0     1.0
Name: Square, dtype: float64

**16. Rank a Series**

In [38]:
# recall the marks_series
marks_series.rank( ascending=True, pct=False)

Maths             1.0
Science           4.0
Art and Craft     2.0
Social Science    3.0
dtype: float64

In [39]:
# recall the square series
square.sort_index(ascending = True)

1.0     1.0
2.0     4.0
2.0     4.0
3.0     9.0
3.0     9.0
5.0    25.0
6.0    36.0
6.0    36.0
6.0    36.0
7.0    49.0
7.0    49.0
8.0    64.0
9.0    81.0
9.0    81.0
9.0    81.0
Name: Square, dtype: float64

<a id="dataframes"> </a>
## 4. Pandas DataFrames

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> A DataFrame is a tabular representation of data containing an ordered collection of columns, each of which can be a different type (numeric, string, boolean, and so on). <br><br>
                        The DataFrame has both a row and column index; it can be thought of as a dict of Series all sharing the same index. In a data frame, the data is stored as one or more two-dimensional blocks rather than a list, dict, or some other collection of one-dimensional arrays.
<br><br>
                        While a DataFrame is physically two-dimensional, it can be used to represent higher dimensional data in a tabular format using hierarchical indexing
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

<a id="creatingDF"> </a>
### 4.1 Creating DataFrames

**1. Creating a data frame a dictionary**

In [40]:
data = {'Subject': ['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art'],
        'Marks': (45, 65, 78, 65, 80, 78),
        'CGPA': [2.5, 3.0, 3.5, 2.0, 4.0, 4.0]}

df = pd.DataFrame(data)
print(df)

    Subject  Marks  CGPA
0     Maths     45   2.5
1   History     65   3.0
2   Science     78   3.5
3   English     65   2.0
4  Georaphy     80   4.0
5       Art     78   4.0


**Note:** Like Series, the resulting DataFrame is assigned index automatically. And the 'Marks' values are in a tuple.

**2. Another way to create dataframe from dictionary**

In [41]:
data = [{'Subject': 'Maths', 'Marks': 45, 'CGPA':2.5},
        {'Subject':'History', 'Marks':65, 'CGPA':3.0},
        {'Subject':'Science', 'Marks':78, 'CGPA':3.5},
        {'Subject':'English', 'Marks':65, 'CGPA':2.0},
        {'Subject':'Georaphy', 'Marks':80, 'CGPA':4.0},
        {'Subject':'Art', 'Marks':78, 'CGPA':4.0}]

df = pd.DataFrame(data)
print(df)

    Subject  Marks  CGPA
0     Maths     45   2.5
1   History     65   3.0
2   Science     78   3.5
3   English     65   2.0
4  Georaphy     80   4.0
5       Art     78   4.0


**3. To create dataframe from series**

In [43]:
Subject = pd.Series(['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art'])
Marks = pd.Series([45, 65, 78, 65, 80, 78])
CGPA = pd.Series([2.5, 3.0, 3.5, 2.0, 4.0, 4.0])

In [44]:
pd.DataFrame([Subject,Marks,CGPA], index = ['Subject','Marks','CGPA'])

0        1        2        3         4    5
Subject  Maths  History  Science  English  Georaphy  Art
Marks       45       65       78       65        80   78
CGPA       2.5      3.0      3.5      2.0       4.0  4.0

However to want a vertical dataframe so we use `.T`. The 'T' stands for transpose.

In [45]:
pd.DataFrame([Subject,Marks,CGPA], index = ['Subject','Marks','CGPA']).T

Subject Marks CGPA
0     Maths    45  2.5
1   History    65  3.0
2   Science    78  3.5
3   English    65  2.0
4  Georaphy    80  4.0
5       Art    78  4.0

**Remark:** Assign a name to the data frame and then use `.T` to transpose it.

**4. To create dataframe from lists**

In [46]:
Subject = ['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art']
Marks = [45, 65, 78, 65, 80, 78]
CGPA = [2.5, 3.0, 3.5, 2.0, 4.0, 4.0]

In [ ]:
pd.DataFrame([Subject,Marks,CGPA], index = ['Subject','Marks','CGPA']).T

Subject Marks CGPA
0     Maths    45  2.5
1   History    65  3.0
2   Science    78  3.5
3   English    65  2.0
4  Georaphy    80  4.0
5       Art    78  4.0

**5. To read data from csv file**

In [47]:
data = pd.read_csv("example.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'example.csv'

In [48]:
type(data)

list

On checking the data type, we notice it is read as pandas data frame.

In [49]:
print(data)

[{'Subject': 'Maths', 'Marks': 45, 'CGPA': 2.5}, {'Subject': 'History', 'Marks': 65, 'CGPA': 3.0}, {'Subject': 'Science', 'Marks': 78, 'CGPA': 3.5}, {'Subject': 'English', 'Marks': 65, 'CGPA': 2.0}, {'Subject': 'Georaphy', 'Marks': 80, 'CGPA': 4.0}, {'Subject': 'Art', 'Marks': 78, 'CGPA': 4.0}]


**6. To print head of the data**

In [50]:
data.head()

AttributeError: 'list' object has no attribute 'head'

By default, the `.head()` will display **first** five rows. However, we can set the desired number of rows to be displayed.

Say we want to see the first 9 rows, we write the number 9 in the parentheses.

In [ ]:
data.head(9)

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32
5   21              43           1.52
6   10              32           1.65
7   57              34           1.61
8   75              23           1.24

**7. To print tail of the data**

In [ ]:
data.tail()

Age  Weight (in kg)  Height (in m)
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

By default, the `.tail()` will display **last** five rows. However, we can set the desired number of rows to be displayed.

Say we want to see the last 14 rows, we write the number 14 in the parentheses.

In [ ]:
data.tail(14)

Age  Weight (in kg)  Height (in m)
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

**8. To obtain the dimension of the data**

In [ ]:
data.shape

(23, 3)

**9. To know the data types of a data frame**

In [ ]:
data.dtypes

Age                 int64
Weight (in kg)      int64
Height (in m)     float64
dtype: object

We see the data type of each variable.

**10. To know some information of the data**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             23 non-null     int64  
 1   Weight (in kg)  23 non-null     int64  
 2   Height (in m)   23 non-null     float64
dtypes: float64(1), int64(2)
memory usage: 684.0 bytes


We see this output gives the number of rows present in the data `RangeIndex: 23 entries, 0 to 22` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 3 columns)`.

Consider `Age 23 non-null int64` indicates that the column named 'Age' has 23 non-null observations having the data type 'int64'

And finally the memory used to save this dataframe is 680 bytes.

**11. To check the data type of column in the data frame**

In [ ]:
type(data.Age)

pandas.core.series.Series

In [ ]:
type(data["Weight (in kg)"])

pandas.core.series.Series

In [ ]:
type(data["Height (in m)"])

pandas.core.series.Series

**Note that every column of the data frame is a pandas Series.**

<a id="manipulatingDF"> </a>
### 4.2  Manipulating DataFrames

### Add new column and rows

<table align="left">
    <tr>
        <td width="8%">
            <img src="caution.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> CAUTION:<br>
                        1. DataFrame[column] works for any column name, but DataFrame.column only works when the column name is a valid Python variable name.<br>
                        2. New columns cannot be created with the ` data.BMI ` syntax.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

**1. Adding a new column to the data set**

In [ ]:
data["BMI"] = data["Weight (in kg)"] / data["Height (in m)"]**2

In [ ]:
data

Age  Weight (in kg)  Height (in m)        BMI
0    45              60           1.35  32.921811
1    12              43           1.21  29.369579
2    54              78           1.50  34.666667
3    26              65           1.21  44.395875
4    68              50           1.32  28.696051
5    21              43           1.52  18.611496
6    10              32           1.65  11.753903
7    57              34           1.61  13.116778
8    75              23           1.24  14.958377
9    32              21           1.52   9.089335
10   23              53           1.50  23.555556
11   34              65           1.76  20.983988
12   55              89           1.65  32.690542
13   23              45           1.75  14.693878
14   56              76           1.69  26.609713
15   67              78           1.85  22.790358
16   26              65           1.21  44.395875
17   56              74           1.69  25.909457
18   67              78           1.85  22.790358
19   26              65           1.21  44.395875
20   68              50           1.32  28.696051
21   56              76           1.69  26.609713
22   67              78           1.85  22.790358

In [ ]:
data.shape

(23, 4)

**2. Adding a new row to the data set**

A new row can be added using the function copy()

In [ ]:
data_copy = data.copy()
data_copy.loc[23] = [45, 85, 1.8, 26.3]

In [ ]:
data_copy

Age  Weight (in kg)  Height (in m)        BMI
0   45.0            60.0           1.35  32.921811
1   12.0            43.0           1.21  29.369579
2   54.0            78.0           1.50  34.666667
3   26.0            65.0           1.21  44.395875
4   68.0            50.0           1.32  28.696051
5   21.0            43.0           1.52  18.611496
6   10.0            32.0           1.65  11.753903
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
9   32.0            21.0           1.52   9.089335
10  23.0            53.0           1.50  23.555556
11  34.0            65.0           1.76  20.983988
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
16  26.0            65.0           1.21  44.395875
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
19  26.0            65.0           1.21  44.395875
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358
23  45.0            85.0           1.80  26.300000

We see that a new column number 23 has be added to the data.

**3. Indexing a dataframe using `.iloc`**

`DataFrame.iloc[]` method is used when the index label of a data frame is something other than numeric series of 0, 1, 2, 3….n or in case the user doesn’t know the index label.

We shall work on the BMI data set.

#### Select the 2nd row

In [ ]:
data.iloc[2]

Age               54.000000
Weight (in kg)    78.000000
Height (in m)      1.500000
BMI               34.666667
Name: 2, dtype: float64

#### Select 4th, 7th and 10th rows

In [ ]:
data.iloc[[4,7,10]]

Age  Weight (in kg)  Height (in m)        BMI
4    68              50           1.32  28.696051
7    57              34           1.61  13.116778
10   23              53           1.50  23.555556

We use two square brackets since we are passing a list of row numbers to be accessed.

#### Select 12th to 17th rows

In [ ]:
data.iloc[12:17]

Age  Weight (in kg)  Height (in m)        BMI
12   55              89           1.65  32.690542
13   23              45           1.75  14.693878
14   56              76           1.69  26.609713
15   67              78           1.85  22.790358
16   26              65           1.21  44.395875

#### Select the 1st column

In [ ]:
data.iloc[:, 1]

0     60
1     43
2     78
3     65
4     50
5     43
6     32
7     34
8     23
9     21
10    53
11    65
12    89
13    45
14    76
15    78
16    65
17    74
18    78
19    65
20    50
21    76
22    78
Name: Weight (in kg), dtype: int64

#### Select the last column

In [ ]:
data.iloc[:,-1]

0     32.921811
1     29.369579
2     34.666667
3     44.395875
4     28.696051
5     18.611496
6     11.753903
7     13.116778
8     14.958377
9      9.089335
10    23.555556
11    20.983988
12    32.690542
13    14.693878
14    26.609713
15    22.790358
16    44.395875
17    25.909457
18    22.790358
19    44.395875
20    28.696051
21    26.609713
22    22.790358
Name: BMI, dtype: float64

To select the last column we use -1, to select the second last column we use -2

#### Select the first two columns

In [ ]:
data.iloc[:,0:2]

Age  Weight (in kg)
0    45              60
1    12              43
2    54              78
3    26              65
4    68              50
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53
11   34              65
12   55              89
13   23              45
14   56              76
15   67              78
16   26              65
17   56              74
18   67              78
19   26              65
20   68              50
21   56              76
22   67              78

#### Select the first two columns and 5 to 10 rows

In [ ]:
data.iloc[5:11, 0:2]

Age  Weight (in kg)
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53

**4. Indexing a dataframe using `.loc`**

`DataFrame.loc[]` method is a method that takes only index labels and returns row or dataframe if the index label exists in the caller data frame. <br>
`DataFrame.loc[Row_names, column_names]` is used to select or index rows or columns based on their name.

#### Select 1 to 5 rows and 2nd and 4th columns

In [ ]:
data.loc[1:5,["Weight (in kg)","BMI"]]

Weight (in kg)        BMI
1              43  29.369579
2              78  34.666667
3              65  44.395875
4              50  28.696051
5              43  18.611496

**Note:** the row names are numbers

**5. Selecting columns by specifying column names**

#### Select the column 'Age'

In [ ]:
data.Age

0     45
1     12
2     54
3     26
4     68
5     21
6     10
7     57
8     75
9     32
10    23
11    34
12    55
13    23
14    56
15    67
16    26
17    56
18    67
19    26
20    68
21    56
22    67
Name: Age, dtype: int64

**Remark:** Using this method we can select only one column.

In [ ]:
# OR
data["Age"]

0     45
1     12
2     54
3     26
4     68
5     21
6     10
7     57
8     75
9     32
10    23
11    34
12    55
13    23
14    56
15    67
16    26
17    56
18    67
19    26
20    68
21    56
22    67
Name: Age, dtype: int64

#### Select the column 'Age' and 'BMI'

In [ ]:
data[["Age","BMI"]]

Age        BMI
0    45  32.921811
1    12  29.369579
2    54  34.666667
3    26  44.395875
4    68  28.696051
5    21  18.611496
6    10  11.753903
7    57  13.116778
8    75  14.958377
9    32   9.089335
10   23  23.555556
11   34  20.983988
12   55  32.690542
13   23  14.693878
14   56  26.609713
15   67  22.790358
16   26  44.395875
17   56  25.909457
18   67  22.790358
19   26  44.395875
20   68  28.696051
21   56  26.609713
22   67  22.790358

**6. Sort the data frame on the basis of values in a column **

Each column of a pandas DataFrame is treated as a pandas Series. The `.sort_values()` in DataFrames works similar to the `pandas.Series`

In [ ]:
# print head() of 'data'
data.head()

Age  Weight (in kg)  Height (in m)        BMI
0   45              60           1.35  32.921811
1   12              43           1.21  29.369579
2   54              78           1.50  34.666667
3   26              65           1.21  44.395875
4   68              50           1.32  28.696051

In [ ]:
# sort the data frame on basis of 'Age' values
# by default the values will get sorted in ascending order
data.sort_values('Age')

#Note: 'ascending = False' will sort the data frame in descending order

Age  Weight (in kg)  Height (in m)        BMI
6    10              32           1.65  11.753903
1    12              43           1.21  29.369579
5    21              43           1.52  18.611496
13   23              45           1.75  14.693878
10   23              53           1.50  23.555556
19   26              65           1.21  44.395875
3    26              65           1.21  44.395875
16   26              65           1.21  44.395875
9    32              21           1.52   9.089335
11   34              65           1.76  20.983988
0    45              60           1.35  32.921811
2    54              78           1.50  34.666667
12   55              89           1.65  32.690542
21   56              76           1.69  26.609713
14   56              76           1.69  26.609713
17   56              74           1.69  25.909457
7    57              34           1.61  13.116778
15   67              78           1.85  22.790358
18   67              78           1.85  22.790358
22   67              78           1.85  22.790358
4    68              50           1.32  28.696051
20   68              50           1.32  28.696051
8    75              23           1.24  14.958377

**7. Rank the dataframe**

In [ ]:
# rank the data frame 'data' in descending order based on 'BMI'
# 'method = min' assigns the minimum rank to highest equal value of 'BMI'
data['BMI_ranked'] = data['BMI'].rank(ascending = 0, method  = 'min')
data

Age  Weight (in kg)  Height (in m)        BMI  BMI_ranked
0    45              60           1.35  32.921811         5.0
1    12              43           1.21  29.369579         7.0
2    54              78           1.50  34.666667         4.0
3    26              65           1.21  44.395875         1.0
4    68              50           1.32  28.696051         8.0
5    21              43           1.52  18.611496        18.0
6    10              32           1.65  11.753903        22.0
7    57              34           1.61  13.116778        21.0
8    75              23           1.24  14.958377        19.0
9    32              21           1.52   9.089335        23.0
10   23              53           1.50  23.555556        13.0
11   34              65           1.76  20.983988        17.0
12   55              89           1.65  32.690542         6.0
13   23              45           1.75  14.693878        20.0
14   56              76           1.69  26.609713        10.0
15   67              78           1.85  22.790358        14.0
16   26              65           1.21  44.395875         1.0
17   56              74           1.69  25.909457        12.0
18   67              78           1.85  22.790358        14.0
19   26              65           1.21  44.395875         1.0
20   68              50           1.32  28.696051         8.0
21   56              76           1.69  26.609713        10.0
22   67              78           1.85  22.790358        14.0

From the above data frame, we can see that 'BMI = 44.395875' is repeating thrice; thus the method = 'min' will assign the minimum rank (=1) to all the three values of BMI. The rank '4' will be assigned to the second largest value of BMI and so on. Thus, there is no rank equal to 2 and 3.

In [ ]:
# method = 'dense' assigns same rank to all the same BMI values
data['BMI_densed_rank'] = data['BMI'].rank(method = 'dense')
data

Age  Weight (in kg)  Height (in m)        BMI  BMI_ranked  BMI_densed_rank
0    45              60           1.35  32.921811         5.0             15.0
1    12              43           1.21  29.369579         7.0             13.0
2    54              78           1.50  34.666667         4.0             16.0
3    26              65           1.21  44.395875         1.0             17.0
4    68              50           1.32  28.696051         8.0             12.0
5    21              43           1.52  18.611496        18.0              6.0
6    10              32           1.65  11.753903        22.0              2.0
7    57              34           1.61  13.116778        21.0              3.0
8    75              23           1.24  14.958377        19.0              5.0
9    32              21           1.52   9.089335        23.0              1.0
10   23              53           1.50  23.555556        13.0              9.0
11   34              65           1.76  20.983988        17.0              7.0
12   55              89           1.65  32.690542         6.0             14.0
13   23              45           1.75  14.693878        20.0              4.0
14   56              76           1.69  26.609713        10.0             11.0
15   67              78           1.85  22.790358        14.0              8.0
16   26              65           1.21  44.395875         1.0             17.0
17   56              74           1.69  25.909457        12.0             10.0
18   67              78           1.85  22.790358        14.0              8.0
19   26              65           1.21  44.395875         1.0             17.0
20   68              50           1.32  28.696051         8.0             12.0
21   56              76           1.69  26.609713        10.0             11.0
22   67              78           1.85  22.790358        14.0              8.0

Here, dense method assigns minimum rank (=1) to minimum value (=9.089335) of the BMI. Rank 2 will be assigned to BMI value greater than min=9.089335 and so on. Thus, no rank is skipped in the dense method.

**8. To check for missing values**

We shall import a new dataset.

In [ ]:
missing_data = pd.read_csv("example_missingdata.csv")
missing_data

Age  Weight (in kg)  Height (in m)
0   45.0            60.0           1.35
1   12.0            43.0           1.21
2   54.0            78.0           1.50
3   26.0            65.0           1.21
4   68.0            50.0           1.32
5   21.0             NaN           1.52
6   10.0            32.0           1.65
7   57.0            34.0           1.61
8   75.0            23.0           1.24
9   32.0            21.0           1.52
10  23.0             NaN           1.50
11  34.0            65.0           1.76
12  55.0            89.0           1.65
13  23.0            45.0           1.75
14  56.0            76.0           1.69
15  67.0            78.0           1.85
16  26.0            65.0           1.21
17  56.0            74.0           1.69
18  67.0            78.0            NaN
19  26.0            65.0           1.21
20  68.0            50.0           1.32
21   NaN            76.0           1.69
22  67.0            78.0           1.85

In [ ]:
missing_data.isnull().sum()

Age               1
Weight (in kg)    2
Height (in m)     1
dtype: int64

The function `.isnull` check whether the data is missing. The `sum()` sums the number of 'True' values in the column. The final output gives the number of missing values in each column.

Here, we see there are 2 missing values in the 'weight' column and one missing value in other columns.

<a id="reading_data"> </a>
## 5.  Reading Data from Different Sources

**1. Read a `.xlsx` file**

In [ ]:
# in case of error :  install openpyxl
pd.read_excel('example.xlsx')

Age  Gender  Salary
0   45    Male   40000
1   12    Male       0
2   54  Female  150000
3   26    Male   30000
4   64  Female   15000
5   21  Female   25600

**2. Read a `.zip` file**

In [ ]:
import zipfile
with zipfile.ZipFile('data.zip') as z:
    with z.open('example.csv') as f:
        file = pd.read_csv(f)
        print(file.head())

   Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32


**3. Read a `.html` file**

In [ ]:
import pandas as pd
# prerequisites : install lxml
df = pd.read_html('Sheet1.html', header=1, index_col=0)
df

[   Age  Gender  Salary
 1                     
 2   45    Male   40000
 3   12    Male       0
 4   54  Female  150000
 5   26    Male   30000
 6   64  Female   15000
 7   21  Female   25600]

**4. Read a `.txt` file**

In [ ]:
data = pd.read_csv('example.txt', sep="\t")
data.head()

Country  Birth rate  Life expectancy
0      Vietnam       1.822        74.828244
1      Vanuatu       3.869        70.819488
2        Tonga       3.911        72.150659
3  Timor-Leste       5.578        61.999854
4     Thailand       1.579        73.927659

**5. Read a `.json` file**

In [ ]:
pd.read_json('iris.json')

sepalLength  sepalWidth  petalLength  petalWidth    species
0            5.1         3.5          1.4         0.2     setosa
1            4.9         3.0          1.4         0.2     setosa
2            4.7         3.2          1.3         0.2     setosa
3            4.6         3.1          1.5         0.2     setosa
4            5.0         3.6          1.4         0.2     setosa
..           ...         ...          ...         ...        ...
145          6.7         3.0          5.2         2.3  virginica
146          6.3         2.5          5.0         1.9  virginica
147          6.5         3.0          5.2         2.0  virginica
148          6.2         3.4          5.4         2.3  virginica
149          5.9         3.0          5.1         1.8  virginica

[150 rows x 5 columns]

**6. Read a `.xml` file**

In [2]:
 import xml.etree.ElementTree as ET

tree = ET.parse("xml_file.xml")
root = tree.getroot()

df_col = ["Name", "Gender", "Marks"]
rows = []

for node in root:
    name = node.attrib.get("name")
    gender = node.find("gender").text if node is not None else None
    marks = node.find("marks").text if node is not None else None

    rows.append({"Name": name, "Gender": gender,
                 "Marks": marks})

xml_df = pd.DataFrame(rows, columns = df_col)
xml_df

FileNotFoundError: [Errno 2] No such file or directory: 'xml_file.xml'

# New Section